modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java

```diff
@@ -0,0 +1,114 @@
+
+/**
+ * A {@link ReconnectableConnectionProviderWrapper} which makes sure that by the time the
+ * {@link ConnectionProvider#connect()} method is invoked on the delegate, the authorization dance has
+ * been completed and the {@link AuthorizationCodeState} and {@link OAuthCallbackValue} fields have
+ * been properly injected
+ *
+ * @since 4.0
+ */
+public class OAuthConnectionProviderWrapper<C> extends ReconnectableConnectionProviderWrapper<C> implements NoConnectivityTest {
+
+  private final OAuthConfig oauthConfig;
+  private final Map<Field, String> callbackValues;
+  private final ExtensionsOAuthManager oauthManager;
+  private final FieldSetter<ConnectionProvider<C>, AuthorizationCodeState> authCodeStateSetter;
+  private final Once dance;
+
+  public OAuthConnectionProviderWrapper(ConnectionProvider<C> delegate,
+                                        OAuthConfig oauthConfig,
+                                        Map<Field, String> callbackValues,
+                                        ExtensionsOAuthManager oauthManager,
+                                        boolean disableValidation,
+                                        RetryPolicyTemplate retryPolicyTemplate) {
+    super(delegate, disableValidation, retryPolicyTemplate);
+    this.oauthConfig = oauthConfig;
+    this.oauthManager = oauthManager;
+    authCodeStateSetter = getAuthCodeStateSetter(delegate);
+    this.callbackValues = unmodifiableMap(callbackValues);
+    dance = Once.of(this::updateAuthState);
+  }
+
+  @Override
+  public C connect() throws ConnectionException {
+    dance.runOnce();
+    return super.connect();
+  }
+
+  public void updateAuthState() {
+    ResourceOwnerOAuthContext context = getContext();
+
+    final ConnectionProvider<C> delegate = getDelegate();
+    authCodeStateSetter.set(delegate, toAuthorizationCodeState(context));
+
+    Map<String, Object> responseParameters = context.getTokenResponseParameters();
+    callbackValues.keySet().forEach(field -> {
+      String key = field.getName();
+      if (responseParameters.containsKey(key)) {
+        new FieldSetter<>(field).set(delegate, responseParameters.get(key));
+      }
+    });
+  }
+
+  public String getResourceOwnerId() {
+    return getContext().getResourceOwnerId();
+  }
+
+  private FieldSetter<ConnectionProvider<C>, AuthorizationCodeState> getAuthCodeStateSetter(ConnectionProvider<C> delegate) {
+    List<Field> stateFields = getFields(delegate.getClass()).stream()
+        .filter(f -> f.getType().equals(AuthorizationCodeState.class))
+        .collect(toList());
+
+    if (stateFields.size() != 1) {
+      throw new IllegalConnectionProviderModelDefinitionException(
+                                                                  format("Connection Provider of class '%s' uses OAuth2 authorization code grant type and thus should contain "
+                                                                      + "one (and only one) field of type %s. %d were found",
```

Comment:
this is an exclusive concern of the Java version of it. Expensive and not all correct to do it in a generic model validator